src/uu/numfmt/src/numfmt.rs

```diff
@@ -28,36 +30,53 @@ const USAGE: &str = help_usage!("numfmt.md");
 
 fn handle_args<'a>(args: impl Iterator<Item = &'a str>, options: &NumfmtOptions) -> UResult<()> {
     for l in args {
-        match format_and_print(l, options) {
-            Ok(_) => Ok(()),
-            Err(e) => Err(NumfmtError::FormattingError(e.to_string())),
-        }?;
+        format_and_handle_validation(l, options)?;
     }
-
     Ok(())
 }
 
 fn handle_buffer<R>(input: R, options: &NumfmtOptions) -> UResult<()>
 where
     R: BufRead,
 {
-    let mut lines = input.lines();
-    for (idx, line) in lines.by_ref().enumerate() {
-        match line {
-            Ok(l) if idx < options.header => {
-                println!("{l}");
+    for (idx, line_result) in input.lines().by_ref().enumerate() {
+        match line_result {
+            Ok(line) if idx < options.header => {
+                println!("{line}");
                 Ok(())
             }
-            Ok(l) => match format_and_print(&l, options) {
-                Ok(_) => Ok(()),
-                Err(e) => Err(NumfmtError::FormattingError(e.to_string())),
-            },
-            Err(e) => Err(NumfmtError::IoError(e.to_string())),
+            Ok(line) => format_and_handle_validation(line.as_ref(), options),
+            Err(err) => return Err(Box::new(NumfmtError::IoError(err.to_string()))),
         }?;
     }
     Ok(())
 }
 
+fn format_and_handle_validation(input_line: &str, options: &NumfmtOptions) -> UResult<()> {
+    let handled_line = format_and_print(input_line, options);
+
+    if let Err(error_message) = handled_line {
+        match options.invalid {
+            InvalidModes::Abort => {
+                return Err(Box::new(NumfmtError::FormattingError(error_message)));
+            }
+            InvalidModes::Fail => {
+                show!(NumfmtError::FormattingError(error_message));
+                println!("{input_line}");
+            }
+            InvalidModes::Ignore => {
+                println!("{input_line}");
+            }
+            InvalidModes::Warn => {
+                show_error!("{}", error_message);
+                println!("{input_line}");
+            }
+        };
+    }
+
+    Ok(())
+}
```

Comment:
what about something like ?

```suggestion
fn format_and_handle_validation(input_line: &str, options: &NumfmtOptions) -> UResult<()> {
    let handled_line = format_and_print(input_line, options);

    if let Err(error_message) = handled_line {
        match options.invalid {
            InvalidModes::Abort => {
                return Err(Box::new(NumfmtError::FormattingError(error_message)));
            }
            InvalidModes::Fail => {
                show!(NumfmtError::FormattingError(error_message));
            }
            InvalidModes::Warn => {
                show_error!("{}", error_message);
            }
            InvalidModes::Ignore => {}
        };
        println!("{}", input_line);
    }

    Ok(())
}

```